### DISCUSSION TITLE
Cumulative report variables

### DISCUSSION THREAD

**Discussion by LenLon** [APSIMInitiative/ApsimX#9308] [Q&A] [answered]
I want to report cumulative variables such as total Nitrogen and WaterUptake over the growing season. I've seen some other issues mentioning this is only really doable using manager scripts, and I have this:

```
using Models.Core;
using Models.Interfaces;
using Models.PMF;
using APSIM.Shared.Utilities;
using System;

namespace Models
{
    [Serializable] 
    public class Script : Model
    {
    	[Link] private Clock Clock;
    	
        [Description("Crop")]
        public Plant Crop { get; set; }
        
        public double total_N_uptake { get; set; }
        
        public double total_water_uptake { get; set; }
        
        [EventSubscribe("DoManagement")]
        private void DoDailyCalculations(object sender, EventArgs e)
        {
        	if (Crop.IsAlive)
        	{
        		total_N_uptake += MathUtilities.Sum(Crop.NitrogenUptake);
        		total_water_uptake += MathUtilities.Sum(Crop.WaterUptake);
        	}
        	
        	else
        	{
        		total_N_uptake = 0;
        		total_water_uptake = 0;
        	}
        }
    }
}
```

This throws, giving this helpful error message:

```
ERROR in file: C:\Users\Lenlon\Documents\Uni\PhD\Papers\WestAfrica_Maize_ML\sims\WestAfrica_C4Maize.apsimx
Simulation name: Simulation
Models.Core.SimulationException
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at APSIM.Shared.Utilities.MathUtilities.Multiply_Value(Double[] value1, Double value2) in C:\Users\ContainerAdministrator\AppData\Local\Temp\ApsimX\APSIM.Shared\Utilities\MathUtilities.cs:line 134
   at Models.ScriptSimulationsSimulationFieldCropStatus11.DoDailyCalculations(Object sender, EventArgs e)
   at Models.Clock.OnDoCommence(Object _, CommenceArgs e) in C:\Users\ContainerAdministrator\AppData\Local\Temp\ApsimX\Models\Clock.cs:line 359
   at Models.Core.Simulation.Run(CancellationTokenSource cancelToken) in C:\Users\ContainerAdministrator\AppData\Local\Temp\ApsimX\Models\Core\Simulation.cs:line 326
   --- End of inner exception stack trace ---
   at Models.Core.Simulation.Run(CancellationTokenSource cancelToken) in C:\Users\ContainerAdministrator\AppData\Local\Temp\ApsimX\Models\Core\Simulation.cs:line 326
   at APSIM.Shared.JobRunning.JobRunner.RunActualJob(IRunnable job, IJobManager jobManager) in C:\Users\ContainerAdministrator\AppData\Local\Temp\ApsimX\APSIM.Shared\JobRunning\JobRunner.cs:line 214
```

I assume it's because I may not just sum up the read-only lists that are NitrogenUptake and WaterUptake, but how would I go about it? I want the cumulative sum of all layers, so to speak. You can get the sum over all layers using `Sum()` in the report module, but here this does not seem to work.

**Comment by hol353** [comment]
The error mentions ```MathUtilities.Multiply_Value```. Your script doesn't call this method so the error must be coming from another script in your simulation.

Do one of these outputs work?

```
sum of sum(Crop.NitrogenUptake) from Plant.Sowing to [Clock].Today as NUptake
```
or
```
sum of sum(Crop.NitrogenUptake) from Plant.Sowing to Plant.Harvesting as NUptake
```

This should be cumulative if you output daily.

**Reply by LenLon** [reply]
Thank you for the suggestions!
The error does refer to "CropStatus11" as the script, which is the one shown above. None of my other scripts use MathUtilities.Multiply either..

Inputting your suggestions in the report module yields 
```
Error in report Report: Invalid report variables found:
WaterUptake: Variable [C4Maize].WaterUptake evaluated to NULL in expression: Sum([C4Maize].WaterUptake)
```
I do NOT output daily though, using the routine from my other discussion thread I output at each phenological stage. So it might throw when trying to report NitrogenUptake at the Sowing stage?

**Reply by LenLon** [reply, marked as answer]
I managed to fix it with the same suggestion you made in the other discussion thread, the problem was the mismatch between the Event `DoManagement` and the function name `DoDailyCalculations` OR the event not being `DoManagementCalculations` (?). Setting the event and function to `DoManagementCalculations` fixed the problem, giving me cumulative values by phenological stage, so great success!

**Comment by sno036** [comment]
Note that you can also do seasonal cumulative reporting just using Report - no Manager needed. See the Report tutorial.
